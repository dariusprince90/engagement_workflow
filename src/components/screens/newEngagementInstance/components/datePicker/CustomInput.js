import { forwardRef, memo } from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

const propTypes = {
  onClick: PropTypes.func
};

// - we're using the whole props object here instead of destructuring into separate vars
// - these props are passed from the react-datepicker
// - we want to ensure we have all the props and thus, shouldn't destructure
let CustomInput = forwardRef((props, ref) => {
  // **********************************************************************
  // * constants

  // **********************************************************************
  // * component vars

  // **********************************************************************
  // * functions

  // **********************************************************************
  // * side effects

  // **********************************************************************
  // * render

  return (
    <div className="input-group">
      <input {...props} ref={ref} />
      <div className="input-group-append">
        <span className="input-group-text icon-clickable" onClick={props.onClick}>
          <FontAwesomeIcon icon="fa-solid fa-calendar-days" />
        </span>
      </div>
    </div>
  );
});

CustomInput = memo(CustomInput);
CustomInput.propTypes = propTypes;
CustomInput.displayName = 'CustomInput';

export default CustomInput;
