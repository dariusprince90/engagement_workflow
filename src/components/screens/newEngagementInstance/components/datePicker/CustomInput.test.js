/* eslint-disable react/prop-types */

import ReactDOM from 'react-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import CustomInput from './CustomInput';

// **********************************************************************
// * constants

const testIds = {
  fontAwesomeIcon: 'font-awesome-icon'
};

const defaultProps = {
  onClick: jest.fn()
};

// **********************************************************************
// * functions

const getComponentToRender = (props) => {
  return <CustomInput {...props} />;
};

// **********************************************************************
// * mock external dependencies

jest.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon: ({ icon }) => {
    const props = { icon };
    return <fake-font-awesome-icon {...props} data-testid={testIds.fontAwesomeIcon} />;
  }
}));

// **********************************************************************
// * unit tests

describe('CustomInput', () => {
  // **********************************************************************
  // * setup

  // **********************************************************************
  // * tear-down

  // **********************************************************************
  // * execution

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(getComponentToRender(defaultProps), div);
  });

  it('renders a proper input group with a calendar icon appended to the input', () => {
    const tree = renderer.create(getComponentToRender(defaultProps)).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('invokes onClick when the icon is clicked', () => {
    const onClick = jest.fn();
    const props = { ...defaultProps, onClick };
    render(getComponentToRender(props));
    expect(onClick).not.toHaveBeenCalled();
    fireEvent.click(screen.getByTestId(testIds.fontAwesomeIcon));
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
