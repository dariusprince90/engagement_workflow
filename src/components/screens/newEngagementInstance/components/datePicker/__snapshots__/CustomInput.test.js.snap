// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomInput renders a proper input group with a calendar icon appended to the input 1`] = `
<div
  className="input-group"
>
  <input
    onClick={[MockFunction]}
  />
  <div
    className="input-group-append"
  >
    <span
      className="input-group-text icon-clickable"
      onClick={[MockFunction]}
    >
      <fake-font-awesome-icon
        data-testid="font-awesome-icon"
        icon="fa-solid fa-calendar-days"
      />
    </span>
  </div>
</div>
`;
